File: ios/Trace/Geometry.swift
import CoreGraphics
import Foundation

// Path sampling and corridor geometry — the Swift counterpart of the web
// version's geometry module.
//
// Shapes are authored in a 0..100 square (y down). A shape is a list of
// strokes, each a mini-SVG-ish string sampled into a polyline:
//
//   M x y            move to
//   L x y            line to
//   Q cx cy x y      quadratic bezier
//   C c1x c1y c2x c2y x y   cubic bezier
//   E cx cy rx ry a0 a1     elliptical arc, degrees, y down
//   S cx cy r0 r1 a0 a1     spiral, radius eased from r0 to r1
//
// Everything downstream works on the sampled polyline, so we never have to do
// real path offsetting.

struct Pt: Equatable {
    var x: CGFloat
    var y: CGFloat
}

@inline(__always) private func lerp(_ a: CGFloat, _ b: CGFloat, _ t: CGFloat) -> CGFloat {
    a + (b - a) * t
}

@inline(__always) private func ellipsePoint(
    _ cx: CGFloat, _ cy: CGFloat, _ rx: CGFloat, _ ry: CGFloat, _ deg: CGFloat
) -> Pt {
    let a = deg * .pi / 180
    return Pt(x: cx + rx * cos(a), y: cy + ry * sin(a))
}

enum PathDSL {
    static let defaultStep: CGFloat = 0.7

    static func sample(_ d: String, step: CGFloat = defaultStep) -> [Pt] {
        let tokens = d.split(whereSeparator: { $0 == " " || $0 == "," || $0 == "\n" })
        var pts: [Pt] = []
        var cx: CGFloat = 0
        var cy: CGFloat = 0
        var i = 0

        func num() -> CGFloat {
            defer { i += 1 }
            guard i < tokens.count, let v = Double(tokens[i]) else { return 0 }
            return CGFloat(v)
        }
        func push(_ x: CGFloat, _ y: CGFloat) {
            if let last = pts.last, hypot(x - last.x, y - last.y) <= 1e-6 { return }
            pts.append(Pt(x: x, y: y))
        }

        while i < tokens.count {
            let cmd = tokens[i]
            i += 1
            switch cmd {
            case "M":
                cx = num(); cy = num()
                push(cx, cy)

            case "L":
                let x = num(), y = num()
                let n = max(2, Int(ceil(hypot(x - cx, y - cy) / step)))
                for k in 1...n {
                    let t = CGFloat(k) / CGFloat(n)
                    push(lerp(cx, x, t), lerp(cy, y, t))
                }
                cx = x; cy = y

            case "Q":
                let qx = num(), qy = num(), x = num(), y = num()
                let approx = hypot(qx - cx, qy - cy) + hypot(x - qx, y - qy)
                let n = max(4, Int(ceil(approx / step)))
                for k in 1...n {
                    let t = CGFloat(k) / CGFloat(n), mt = 1 - t
                    push(
                        mt * mt * cx + 2 * mt * t * qx + t * t * x,
                        mt * mt * cy + 2 * mt * t * qy + t * t * y)
                }
                cx = x; cy = y

            case "C":
                let c1x = num(), c1y = num(), c2x = num(), c2y = num(), x = num(), y = num()
                let approx = hypot(c1x - cx, c1y - cy) + hypot(c2x - c1x, c2y - c1y) + hypot(x - c2x, y - c2y)
                let n = max(6, Int(ceil(approx / step)))
                for k in 1...n {
                    let t = CGFloat(k) / CGFloat(n), mt = 1 - t
                    push(
                        mt * mt * mt * cx + 3 * mt * mt * t * c1x + 3 * mt * t * t * c2x + t * t * t * x,
                        mt * mt * mt * cy + 3 * mt * mt * t * c1y + 3 * mt * t * t * c2y + t * t * t * y)
                }
                cx = x; cy = y

            case "E":
                let ecx = num(), ecy = num(), rx = num(), ry = num(), a0 = num(), a1 = num()
                let arcLen = abs(a1 - a0) * .pi / 180 * ((rx + ry) / 2)
                let n = max(12, Int(ceil(arcLen / step)))
                for k in 0...n {
                    let p = ellipsePoint(ecx, ecy, rx, ry, lerp(a0, a1, CGFloat(k) / CGFloat(n)))
                    push(p.x, p.y)
                    cx = p.x; cy = p.y
                }

            case "S":
                let scx = num(), scy = num(), r0 = num(), r1 = num(), a0 = num(), a1 = num()
                let arcLen = abs(a1 - a0) * .pi / 180 * ((r0 + r1) / 2)
                let n = max(24, Int(ceil(arcLen / step)))
                for k in 0...n {
                    let t = CGFloat(k) / CGFloat(n)
                    let r = lerp(r0, r1, t)
                    let p = ellipsePoint(scx, scy, r, r, lerp(a0, a1, t))
                    push(p.x, p.y)
                    cx = p.x; cy = p.y
                }

            default:
                assertionFailure("Unknown path command \"\(cmd)\" in: \(d)")
            }
        }
        return pts
    }
}

func polylineLength(_ pts: [Pt]) -> CGFloat {
    guard pts.count > 1 else { return 0 }
    var total: CGFloat = 0
    for i in 1..<pts.count { total += hypot(pts[i].x - pts[i - 1].x, pts[i].y - pts[i - 1].y) }
    return total
}

/// Re-space a polyline so consecutive points sit `spacing` apart.
func resample(_ pts: [Pt], spacing: CGFloat) -> [Pt] {
    guard pts.count > 1, spacing > 0 else { return pts }
    var out = [pts[0]]
    var carry: CGFloat = 0
    for i in 1..<pts.count {
        let a = pts[i - 1], b = pts[i]
        let seg = hypot(b.x - a.x, b.y - a.y)
        if seg == 0 { continue }
        var t = spacing - carry
        while t <= seg {
            out.append(Pt(x: lerp(a.x, b.x, t / seg), y: lerp(a.y, b.y, t / seg)))
            t += spacing
        }
        carry = (carry + seg).truncatingRemainder(dividingBy: spacing)
    }
    if let last = pts.last, let tail = out.last, hypot(tail.x - last.x, tail.y - last.y) > spacing * 0.4 {
        out.append(last)
    }
    return out
}

func boundingBox(_ strokes: [[Pt]]) -> CGRect {
    var minX = CGFloat.greatestFiniteMagnitude, minY = CGFloat.greatestFiniteMagnitude
    var maxX = -CGFloat.greatestFiniteMagnitude, maxY = -CGFloat.greatestFiniteMagnitude
    for pts in strokes {
        for p in pts {
            minX = min(minX, p.x); minY = min(minY, p.y)
            maxX = max(maxX, p.x); maxY = max(maxY, p.y)
        }
    }
    return CGRect(x: minX, y: minY, width: maxX - minX, height: maxY - minY)
}

// MARK: - Trial

struct StrokeGeometry {
    var pts: [Pt]
    var len: CGFloat
    var checkpoints: [Pt]

    var cgPath: CGPath {
        let p = CGMutablePath()
        guard let first = pts.first else { return p }
        p.move(to: CGPoint(x: first.x, y: first.y))
        for q in pts.dropFirst() { p.addLine(to: CGPoint(x: q.x, y: q.y)) }
        return p
    }
}

struct Trial {
    let shape: Shape
    let strokes: [StrokeGeometry]
    let corridorPx: CGFloat
    let corridorUnits: CGFloat
    let totalLen: CGFloat
}

/// Build everything the game needs from a shape definition, in view points.
///
/// `corridorUnits` is the full corridor width in shape units, so 18 means the
/// gap between the two lines is 18% of the shape box.
func buildTrial(shape: Shape, corridorUnits: CGFloat, view: CGRect) -> Trial {
    let raw = shape.strokes.map { PathDSL.sample($0) }
    let box = boundingBox(raw)

    // Fit the shape plus half a corridor of breathing room into the view.
    let pad = corridorUnits / 2 + 3
    let scale = min(view.width / (box.width + pad * 2), view.height / (box.height + pad * 2))
    let ox = view.minX + (view.width - box.width * scale) / 2 - box.minX * scale
    let oy = view.minY + (view.height - box.height * scale) / 2 - box.minY * scale

    let corridorPx = corridorUnits * scale
    let strokes: [StrokeGeometry] = raw.map { pts in
        let screen = pts.map { Pt(x: $0.x * scale + ox, y: $0.y * scale + oy) }
        let len = polylineLength(screen)
        // One checkpoint about every third of a corridor width, clamped so short
        // strokes still get a usable number and long ones don't explode.
        let spacing = min(max(corridorPx * 0.34, 10), max(len / 12, 12))
        return StrokeGeometry(pts: screen, len: len, checkpoints: resample(screen, spacing: spacing))
    }

    return Trial(
        shape: shape,
        strokes: strokes,
        corridorPx: corridorPx,
        corridorUnits: corridorUnits,
        totalLen: strokes.reduce(0) { $0 + $1.len })
}

// MARK: - Corridor mask

/// The corridor rasterised once into a byte grid, so asking "is this point
/// inside the lines?" costs one array lookup instead of a distance-to-path scan
/// on every pencil sample — and the pencil delivers a lot of samples.
final class CorridorMask {
    let width: Int
    let height: Int
    private let grid: [UInt8]

    init?(trial: Trial, size: CGSize) {
        width = max(1, Int(size.width.rounded()))
        height = max(1, Int(size.height.rounded()))
        let bytes = width * height
        guard
            let ctx = CGContext(
                data: nil, width: width, height: height, bitsPerComponent: 8,
                bytesPerRow: width, space: CGColorSpaceCreateDeviceGray(),
                bitmapInfo: CGImageAlphaInfo.none.rawValue)
        else { return nil }

        // Flip so the mask shares the view's top-left origin.
        ctx.translateBy(x: 0, y: CGFloat(height))
        ctx.scaleBy(x: 1, y: -1)
        ctx.setFillColor(gray: 0, alpha: 1)
        ctx.fill(CGRect(x: 0, y: 0, width: width, height: height))
        ctx.setStrokeColor(gray: 1, alpha: 1)
        ctx.setLineWidth(trial.corridorPx)
        ctx.setLineCap(.round)
        ctx.setLineJoin(.round)
        for st in trial.strokes {
            ctx.addPath(st.cgPath)
            ctx.strokePath()
        }

        guard let data = ctx.data else { return nil }
        let buf = data.bindMemory(to: UInt8.self, capacity: bytes)
        var g = [UInt8](repeating: 0, count: bytes)
        for i in 0..<bytes { g[i] = buf[i] > 120 ? 1 : 0 }
        grid = g
    }

    @inline(__always) func contains(_ x: CGFloat, _ y: CGFloat) -> Bool {
        let ix = Int(x), iy = Int(y)
        guard ix >= 0, iy >= 0, ix < width, iy < height else { return false }
        return grid[iy * width + ix] == 1
    }
}

File: src/geometry.js
// Path sampling + corridor geometry.
//
// Shapes are authored in a 0..100 square ("shape units"). A shape is a list of
// strokes; each stroke is a mini-SVG-ish string that we sample into a polyline:
//
//   M x y            move to
//   L x y            line to
//   Q cx cy x y      quadratic bezier
//   C c1x c1y c2x c2y x y   cubic bezier
//   E cx cy rx ry a0 a1     elliptical arc, angles in degrees, y down
//   S cx cy r0 r1 a0 a1     spiral arc, radius eased from r0 to r1 over the sweep
//
// Everything downstream (corridor mask, coverage checkpoints, the demo dot)
// works on the sampled polyline, so we never have to do real path offsetting.

const DEFAULT_STEP = 0.7; // shape units between samples

function lerp(a, b, t) {
  return a + (b - a) * t;
}

function ellipsePoint(cx, cy, rx, ry, deg) {
  const a = (deg * Math.PI) / 180;
  return [cx + rx * Math.cos(a), cy + ry * Math.sin(a)];
}

function dist(ax, ay, bx, by) {
  return Math.hypot(bx - ax, by - ay);
}

/** Sample a path string into a flat polyline: [{x, y}, ...] */
export function samplePath(d, step = DEFAULT_STEP) {
  const tok = d.trim().split(/[\s,]+/);
  const pts = [];
  let cx = 0;
  let cy = 0;
  let i = 0;

  const num = () => parseFloat(tok[i++]);
  const push = (x, y) => {
    const last = pts[pts.length - 1];
    if (!last || dist(last.x, last.y, x, y) > 1e-6) pts.push({ x, y });
  };

  while (i < tok.length) {
    const cmd = tok[i++];
    switch (cmd) {
      case 'M': {
        cx = num();
        cy = num();
        push(cx, cy);
        break;
      }
      case 'L': {
        const x = num();
        const y = num();
        const n = Math.max(2, Math.ceil(dist(cx, cy, x, y) / step));
        for (let k = 1; k <= n; k++) push(lerp(cx, x, k / n), lerp(cy, y, k / n));
        cx = x;
        cy = y;
        break;
      }
      case 'Q': {
        const qx = num();
        const qy = num();
        const x = num();
        const y = num();
        const approx = dist(cx, cy, qx, qy) + dist(qx, qy, x, y);
        const n = Math.max(4, Math.ceil(approx / step));
        for (let k = 1; k <= n; k++) {
          const t = k / n;
          const mt = 1 - t;
          push(
            mt * mt * cx + 2 * mt * t * qx + t * t * x,
            mt * mt * cy + 2 * mt * t * qy + t * t * y,
          );
        }
        cx = x;
        cy = y;
        break;
      }
      case 'C': {
        const c1x = num();
        const c1y = num();
        const c2x = num();
        const c2y = num();
        const x = num();
        const y = num();
        const approx =
          dist(cx, cy, c1x, c1y) + dist(c1x, c1y, c2x, c2y) + dist(c2x, c2y, x, y);
        const n = Math.max(6, Math.ceil(approx / step));
        for (let k = 1; k <= n; k++) {
          const t = k / n;
          const mt = 1 - t;
          push(
            mt * mt * mt * cx + 3 * mt * mt * t * c1x + 3 * mt * t * t * c2x + t * t * t * x,
            mt * mt * mt * cy + 3 * mt * mt * t * c1y + 3 * mt * t * t * c2y + t * t * t * y,
          );
        }
        cx = x;
        cy = y;
        break;
      }
      case 'E': {
        const ecx = num();
        const ecy = num();
        const rx = num();
        const ry = num();
        const a0 = num();
        const a1 = num();
        const sweep = Math.abs(a1 - a0);
        const arcLen = ((sweep * Math.PI) / 180) * ((rx + ry) / 2);
        const n = Math.max(12, Math.ceil(arcLen / step));
        for (let k = 0; k <= n; k++) {
          const [x, y] = ellipsePoint(ecx, ecy, rx, ry, lerp(a0, a1, k / n));
          push(x, y);
          cx = x;
          cy = y;
        }
        break;
      }
      case 'S': {
        const scx = num();
        const scy = num();
        const r0 = num();
        const r1 = num();
        const a0 = num();
        const a1 = num();
        const sweep = Math.abs(a1 - a0);
        const arcLen = ((sweep * Math.PI) / 180) * ((r0 + r1) / 2);
        const n = Math.max(24, Math.ceil(arcLen / step));
        for (let k = 0; k <= n; k++) {
          const t = k / n;
          const r = lerp(r0, r1, t);
          const [x, y] = ellipsePoint(scx, scy, r, r, lerp(a0, a1, t));
          push(x, y);
          cx = x;
          cy = y;
        }
        break;
      }
      default:
        throw new Error(`Unknown path command "${cmd}" in: ${d}`);
    }
  }
  return pts;
}

export function polylineLength(pts) {
  let L = 0;
  for (let i = 1; i < pts.length; i++) L += dist(pts[i - 1].x, pts[i - 1].y, pts[i].x, pts[i].y);
  return L;
}

/** Re-space a polyline so consecutive points are `spacing` apart. */
export function resample(pts, spacing) {
  if (pts.length < 2) return pts.slice();
  const out = [pts[0]];
  let carry = 0;
  for (let i = 1; i < pts.length; i++) {
    const a = pts[i - 1];
    const b = pts[i];
    const seg = dist(a.x, a.y, b.x, b.y);
    if (seg === 0) continue;
    let t = spacing - carry;
    while (t <= seg) {
      out.push({ x: lerp(a.x, b.x, t / seg), y: lerp(a.y, b.y, t / seg) });
      t += spacing;
    }
    carry = (carry + seg) % spacing;
  }
  const last = pts[pts.length - 1];
  const tail = out[out.length - 1];
  if (dist(tail.x, tail.y, last.x, last.y) > spacing * 0.4) out.push(last);
  return out;
}

/** Bounding box of every stroke together. */
export function bboxOf(strokes) {
  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;
  for (const pts of strokes) {
    for (const p of pts) {
      if (p.x < minX) minX = p.x;
      if (p.y < minY) minY = p.y;
      if (p.x > maxX) maxX = p.x;
      if (p.y > maxY) maxY = p.y;
    }
  }
  return { minX, minY, maxX, maxY, w: maxX - minX, h: maxY - minY };
}

/**
 * Build everything the game needs from a shape definition, in screen pixels.
 *
 * `corridorUnits` is the full corridor width in shape units (0..100 box), so a
 * value of 18 means the gap between the two lines is 18% of the shape box.
 */
export function buildTrial(shape, corridorUnits, view) {
  const raw = shape.strokes.map((d) => samplePath(d));
  const box = bboxOf(raw);

  // Fit the shape (plus half a corridor of breathing room) into the view.
  const pad = corridorUnits / 2 + 3;
  const spanX = box.w + pad * 2;
  const spanY = box.h + pad * 2;
  const scale = Math.min(view.w / spanX, view.h / spanY);
  const ox = view.x + (view.w - box.w * scale) / 2 - box.minX * scale;
  const oy = view.y + (view.h - box.h * scale) / 2 - box.minY * scale;
  const toScreen = (p) => ({ x: p.x * scale + ox, y: p.y * scale + oy });

  const corridorPx = corridorUnits * scale;
  const strokes = raw.map((pts) => {
    const screen = pts.map(toScreen);
    const len = polylineLength(screen);
    // One checkpoint roughly every third of a corridor width, clamped so short
    // strokes still get a usable number and long ones don't explode.
    const spacing = Math.min(Math.max(corridorPx * 0.34, 10), Math.max(len / 12, 12));
    return { pts: screen, len, checkpoints: resample(screen, spacing) };
  });

  return {
    shape,
    strokes,
    corridorPx,
    scale,
    totalLen: strokes.reduce((s, st) => s + st.len, 0),
  };
}

/**
 * Rasterise the corridor once into a byte grid. Hit-testing a finger position
 * then costs one array lookup instead of a distance-to-polyline scan.
 */
export function buildCorridorMask(trial, w, h) {
  const c = document.createElement('canvas');
  c.width = w;
  c.height = h;
  const ctx = c.getContext('2d', { willReadFrequently: true });
  ctx.strokeStyle = '#fff';
  ctx.lineWidth = trial.corridorPx;
  ctx.lineCap = 'round';
  ctx.lineJoin = 'round';
  for (const st of trial.strokes) {
    ctx.beginPath();
    ctx.moveTo(st.pts[0].x, st.pts[0].y);
    for (let i = 1; i < st.pts.length; i++) ctx.lineTo(st.pts[i].x, st.pts[i].y);
    ctx.stroke();
  }
  const src = ctx.getImageData(0, 0, w, h).data;
  const grid = new Uint8Array(w * h);
  for (let i = 0, p = 3; i < grid.length; i++, p += 4) grid[i] = src[p] > 120 ? 1 : 0;
  return { grid, w, h };
}

export function maskContains(mask, x, y) {
  const ix = x | 0;
  const iy = y | 0;
  if (ix < 0 || iy < 0 || ix >= mask.w || iy >= mask.h) return false;
  return mask.grid[iy * mask.w + ix] === 1;
}
